File: 36. Valid sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {

        const rows = Array.from({ length: 9 }, () => Array(9).fill(false));
        const cols = Array.from({ length: 9 }, () => Array(9).fill(false));
        const boxes = Array.from({ length: 9 }, () => Array(9).fill(false));

        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                const c = board[i][j];
                if (c !== '.') { 
                    const num = parseInt(c) - 1; 
                    const boxIndex = Math.floor(i / 3) * 3 + Math.floor(j / 3); 

                    if (rows[i][num] || cols[j][num] || boxes[boxIndex][num]) {
                        return false; 
                    }
                    
                    rows[i][num] = true;
                    cols[j][num] = true;
                    boxes[boxIndex][num] = true;
                }
            }
        }
        return true; 
    
};

File: 36. Valid Sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        rows = [[False] * 9 for _ in range(9)]
        cols = [[False] * 9 for _ in range(9)]
        boxes = [[False] * 9 for _ in range(9)]

        for i in range(9):
            for j in range(9):
                c = board[i][j]
                if c != '.':  
                    num = int(c) - 1 
                    box_index = (i // 3) * 3 + (j // 3)  

                    if rows[i][num] or cols[j][num] or boxes[box_index][num]:
                        return False  

                    rows[i][num] = True
                    cols[j][num] = True
                    boxes[box_index][num] = True
        return True  